File: sources/doi_resolve.py
"""DOI resolution (no GUI dependencies). Used by the desktop DOI page and the web API."""

from __future__ import annotations

import datetime
import json
import re
from urllib.error import URLError
from urllib.request import Request, urlopen

from .base import PaperMetadata

_ARXIV_DOI_RE = re.compile(
    r"10\.48550/arXiv\.(\d{4}\.\d{4,5}|[a-z\-]+/\d+)", re.IGNORECASE
)

_S2_FIELDS = "title,authors,year,abstract,externalIds,venue,publicationDate,url"


def _strip_doi_url(doi: str) -> str:
    return re.sub(r"^https?://(dx\.)?doi\.org/", "", doi.strip())


def _is_ratelimited(e: Exception) -> bool:
    return "429" in str(e)


def _fetch_url(url: str, timeout: int = 8) -> dict:
    """GET a JSON URL and return parsed dict. Raises on HTTP/network error."""
    req = Request(url, headers={"User-Agent": "linXiv/1.0 (mailto:user@example.com)"})
    with urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read())


def _try_arxiv_doi(doi: str) -> PaperMetadata | None:
    """If doi matches 10.48550/arXiv.ID, fetch directly from arXiv."""
    m = _ARXIV_DOI_RE.search(doi)
    if not m:
        return None
    arxiv_id = m.group(1)
    from .fetch_paper_metadata import fetch_paper_metadata
    from .arxiv_source import _result_to_metadata

    try:
        result = fetch_paper_metadata(arxiv_id)
        return _result_to_metadata(result)
    except Exception as e:
        if _is_ratelimited(e):
            raise ValueError(
                "arXiv rate limit reached. Please wait ~60 s and try again."
            ) from e
        return None


def _try_semantic_scholar(doi: str) -> PaperMetadata | None:
    """
    Look up by DOI on Semantic Scholar.
    If the paper has an arXiv ID, fetch the full arXiv record.
    Otherwise build PaperMetadata from S2 fields.
    """
    try:
        data = _fetch_url(
            f"https://api.semanticscholar.org/graph/v1/paper/DOI:{doi}"
            f"?fields={_S2_FIELDS}"
        )
    except (URLError, json.JSONDecodeError, Exception):
        return None

    if not data or "title" not in data:
        return None

    arxiv_id = (data.get("externalIds") or {}).get("ArXiv")
    if arxiv_id:
        from .fetch_paper_metadata import fetch_paper_metadata
        from .arxiv_source import _result_to_metadata

        try:
            result = fetch_paper_metadata(arxiv_id)
            return _result_to_metadata(result)
        except Exception as e:
            if _is_ratelimited(e):
                raise ValueError(
                    "arXiv rate limit reached. Please wait ~60 s and try again."
                ) from e

    pub_date: datetime.date = datetime.date.today()
    raw_date = data.get("publicationDate")
    if raw_date:
        try:
            pub_date = datetime.date.fromisoformat(raw_date)
        except ValueError:
            year = data.get("year")
            if year:
                pub_date = datetime.date(int(year), 1, 1)
    elif data.get("year"):
        pub_date = datetime.date(int(data["year"]), 1, 1)

    authors = [a["name"] for a in (data.get("authors") or []) if a.get("name")]
    venue = data.get("venue") or ""
    s2_url = data.get("url") or f"https://www.semanticscholar.org/paper/{data.get('paperId', '')}"

    return PaperMetadata(
        paper_id=doi,
        version=1,
        title=data["title"],
        authors=authors,
        published=pub_date,
        summary=data.get("abstract") or "",
        category=venue or None,
        doi=doi,
        url=s2_url,
        source="semanticscholar",
    )


def _try_crossref(doi: str) -> PaperMetadata | None:
    """Full CrossRef metadata fetch — returns whatever the registry knows."""
    try:
        data = _fetch_url(f"https://api.crossref.org/works/{doi}")
    except (URLError, json.JSONDecodeError, Exception):
        return None

    msg = data.get("message", {})
    titles = msg.get("title", [])
    if not titles:
        return None

    title = titles[0]

    authors: list[str] = []
    for a in msg.get("author", []):
        given = a.get("given", "")
        family = a.get("family", "")
        name = f"{given} {family}".strip()
        if name:
            authors.append(name)

    pub_date = datetime.date.today()
    dp = msg.get("published", {}).get("date-parts", [[]])
    if dp and dp[0]:
        parts = dp[0]
        try:
            pub_date = datetime.date(
                parts[0],
                parts[1] if len(parts) > 1 else 1,
                parts[2] if len(parts) > 2 else 1,
            )
        except (ValueError, TypeError):
            pass

    abstract = msg.get("abstract") or ""
    abstract = re.sub(r"<[^>]+>", "", abstract).strip()

    journal = (msg.get("container-title") or [""])[0]
    cr_url = msg.get("URL") or f"https://doi.org/{doi}"

    return PaperMetadata(
        paper_id=doi,
        version=1,
        title=title,
        authors=authors,
        published=pub_date,
        summary=abstract,
        category=journal or None,
        doi=doi,
        url=cr_url,
        source="crossref",
    )


def resolve_doi(doi: str) -> PaperMetadata:
    """
    Resolve a DOI to PaperMetadata via three strategies:
      1. arXiv-issued DOI  → fetch directly from arXiv
      2. Semantic Scholar  → resolves any DOI; uses arXiv ID when available
      3. CrossRef          → last resort; broadest DOI coverage
    Raises ValueError with a human-readable message on failure.
    """
    doi = _strip_doi_url(doi)
    if not doi:
        raise ValueError("Please enter a DOI.")

    meta = _try_arxiv_doi(doi)
    if meta:
        return meta

    meta = _try_semantic_scholar(doi)
    if meta:
        return meta

    meta = _try_crossref(doi)
    if meta:
        return meta

    raise ValueError(
        "Could not resolve this DOI.\n"
        "• Check the DOI is correct\n"
        "• The paper may not be indexed by Semantic Scholar or CrossRef\n"
        "• arXiv-hosted papers use DOIs starting with 10.48550/arXiv."
    )


def _resolve_doi(doi: str) -> PaperMetadata:
    """Backward-compatible name used by the PyQt DOI page."""
    return resolve_doi(doi)

File: src-tauri/crates/core/src/sources/doi_resolve.rs
//! doi_resolve — resolve a DOI to `models::PaperMetadata`. Plan §5.4. In order:
//!   1. arXiv-issued DOI (`10.48550/arXiv.<id>`) -> fetch the arXiv record
//!   2. Semantic Scholar  -> any DOI; uses the arXiv id when S2 exposes one
//!   3. CrossRef          -> last resort (reuses `crossref::parse_doi_body`)
//! NON-NEGOTIABLE: an arXiv 429 hit in strategy 1 or 2's arXiv-id branch is
//! RE-RAISED as a user-facing error, never swallowed into the fallback chain.

use chrono::{NaiveDate, Utc};
use reqwest::StatusCode;
use serde_json::Value;

use std::path::Path;

use super::{arxiv, crossref, http};
use crate::error::{CoreError, Result};
use crate::models::{doi_source_id, PaperMetadata};

/// Fields requested from the Semantic Scholar graph API.
const S2_FIELDS: &str = "title,authors,year,abstract,externalIds,venue,publicationDate,url";
const S2_BASE: &str = "https://api.semanticscholar.org/graph/v1/paper/DOI:";
/// Semantic Scholar is reached over exactly one host; the http guard enforces it.
const S2_HOSTS: &[&str] = &["api.semanticscholar.org"];
const RATELIMIT_MSG: &str = "arXiv rate limit reached. Please wait ~60 s and try again.";

// ---------------------------------------------------------------------------
// Pure helpers (sync, unit-tested).
// ---------------------------------------------------------------------------

/// Strip a leading `http(s)://(dx.)doi.org/` and surrounding whitespace.
pub fn strip_doi_url(doi: &str) -> String {
    let t = doi.trim();
    [
        "https://doi.org/",
        "http://doi.org/",
        "https://dx.doi.org/",
        "http://dx.doi.org/",
    ]
    .iter()
    .find_map(|p| t.strip_prefix(p))
    .unwrap_or(t)
    .to_string()
}

/// An arXiv error is a rate-limit iff its message mentions `429`;
/// `http::arxiv_get` surfaces a 429 as exactly that.
fn is_ratelimited(e: &CoreError) -> bool {
    e.to_string().contains("429")
}

/// If `doi` contains an arXiv-issued DOI (`10.48550/arXiv.<id>`, case-insensitive)
/// return the bare arXiv id (new-style `\d{4}\.\d{4,5}` tried first, else
/// old-style `category/number`); else `None`.
pub(crate) fn arxiv_doi_id(doi: &str) -> Option<String> {
    let prefix = "10.48550/arxiv.";
    let pos = doi.to_ascii_lowercase().find(prefix)?;
    let rest = &doi[pos + prefix.len()..];
    let b = rest.as_bytes();

    // new-style: NNNN.NNNN or NNNN.NNNNN (\d{4}\.\d{4,5})
    if b.len() >= 9
        && b[..4].iter().all(u8::is_ascii_digit)
        && b[4] == b'.'
        && b[5..9].iter().all(u8::is_ascii_digit)
    {
        let end = if b.len() >= 10 && b[9].is_ascii_digit() {
            10
        } else {
            9
        };
        return Some(rest[..end].to_string());
    }

    // old-style: [A-Za-z-]+/\d+
    let cat = b
        .iter()
        .take_while(|&&c| c.is_ascii_alphabetic() || c == b'-')
        .count();
    if cat == 0 || b.get(cat) != Some(&b'/') {
        return None;
    }
    let digits = b[cat + 1..]
        .iter()
        .take_while(|c| c.is_ascii_digit())
        .count();
    if digits == 0 {
        return None;
    }
    Some(rest[..cat + 1 + digits].to_string())
}

/// Published date from a Semantic Scholar record: a valid ISO `publicationDate`
/// wins (an invalid one falls to `year`), else `year` -> Jan 1, else today.
fn s2_published(data: &Value) -> NaiveDate {
    let today = Utc::now().date_naive();
    let year_jan1 = || {
        data.get("year")
            .and_then(Value::as_i64)
            .and_then(|y| NaiveDate::from_ymd_opt(y as i32, 1, 1))
    };
    match data.get("publicationDate").and_then(Value::as_str) {
        Some(raw) if !raw.is_empty() => NaiveDate::parse_from_str(raw, "%Y-%m-%d")
            .ok()
            .or_else(year_jan1)
            .unwrap_or(today),
        _ => year_jan1().unwrap_or(today),
    }
}

/// Build `PaperMetadata` from Semantic Scholar's own fields (the non-arXiv path);
/// `None` when the record has no `title`. The arXiv-id branch lives in `try_semantic_scholar`.
fn parse_s2(data: &Value, doi: &str) -> Option<PaperMetadata> {
    let title = data.get("title").and_then(Value::as_str)?;

    let authors = data
        .get("authors")
        .and_then(Value::as_array)
        .map(|arr| {
            arr.iter()
                .filter_map(|a| a.get("name").and_then(Value::as_str))
                .filter(|s| !s.is_empty())
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default();

    let venue = data
        .get("venue")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty());

    let url = data
        .get("url")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
        .unwrap_or_else(|| {
            let pid = data.get("paperId").and_then(Value::as_str).unwrap_or("");
            format!("https://www.semanticscholar.org/paper/{pid}")
        });

    Some(PaperMetadata {
        source_id: doi_source_id(doi),
        version: 1,
        title: title.to_string(),
        authors,
        published: s2_published(data),
        updated: None,
        summary: data
            .get("abstract")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        category: venue.map(str::to_string),
        categories: None,
        doi: Some(doi.to_string()),
        journal_ref: None,
        comment: None,
        url: Some(url),
        tags: None,
        source: Some("semanticscholar".to_string()),
        author_orcids: None,
    })
}

// ---------------------------------------------------------------------------
// Async strategies (against the sources::http + sources::arxiv seams).
// ---------------------------------------------------------------------------

/// Strategy 1: if `doi` is an arXiv-issued DOI, fetch the arXiv record.
/// `Ok(None)` for a non-arXiv DOI or any non-429 arXiv error; an arXiv 429
/// is re-raised as the user-facing rate-limit error.
async fn try_arxiv_doi(doi: &str, data_dir: &Path) -> Result<Option<PaperMetadata>> {
    let Some(id) = arxiv_doi_id(doi) else {
        return Ok(None);
    };
    match arxiv::fetch_by_id(&id, data_dir).await {
        Ok(m) => Ok(Some(m)),
        Err(e) if is_ratelimited(&e) => Err(CoreError::BadRequest(RATELIMIT_MSG.to_string())),
        Err(_) => Ok(None),
    }
}

/// Strategy 2: look the DOI up on Semantic Scholar. If S2 exposes an arXiv id,
/// fetch the full arXiv record (re-raising a 429); otherwise build from S2's own
/// fields. Any fetch/parse failure resolves to `Ok(None)`.
async fn try_semantic_scholar(doi: &str, data_dir: &Path) -> Result<Option<PaperMetadata>> {
    let url = format!("{S2_BASE}{doi}?fields={S2_FIELDS}");
    let resp = match http::get_guarded(&url, S2_HOSTS).await {
        Ok(r) if r.status() == StatusCode::OK => r,
        _ => return Ok(None),
    };
    let Ok(data) = resp.json::<Value>().await else {
        return Ok(None);
    };
    if data.get("title").is_none() {
        return Ok(None);
    }

    let arxiv_id = data
        .get("externalIds")
        .and_then(|e| e.get("ArXiv"))
        .and_then(Value::as_str);
    if let Some(id) = arxiv_id {
        match arxiv::fetch_by_id(id, data_dir).await {
            Ok(m) => return Ok(Some(m)),
            Err(e) if is_ratelimited(&e) => {
                return Err(CoreError::BadRequest(RATELIMIT_MSG.to_string()))
            }
            // Non-429 arXiv error: fall through and build from S2's own fields.
            Err(_) => {}
        }
    }
    Ok(parse_s2(&data, doi))
}

/// Resolve a DOI, trying arXiv -> Semantic Scholar -> CrossRef. `data_dir` (DI)
/// feeds the arXiv cool-down, `mailto` (DI) the CrossRef polite pool. Errors are
/// user-facing: empty input, a propagated arXiv 429, or "could not resolve".
pub async fn resolve_doi(doi: &str, data_dir: &Path, mailto: &str) -> Result<PaperMetadata> {
    let doi = strip_doi_url(doi);
    if doi.is_empty() {
        return Err(CoreError::BadRequest("Please enter a DOI.".to_string()));
    }

    if let Some(m) = try_arxiv_doi(&doi, data_dir).await? {
        return Ok(m);
    }
    if let Some(m) = try_semantic_scholar(&doi, data_dir).await? {
        return Ok(m);
    }
    if let Some(m) = crossref::fetch_by_doi(&doi, mailto).await {
        return Ok(m);
    }

    Err(CoreError::BadRequest(
        "Could not resolve this DOI.\n\
         • Check the DOI is correct\n\
         • The paper may not be indexed by Semantic Scholar or CrossRef\n\
         • arXiv-hosted papers use DOIs starting with 10.48550/arXiv."
            .to_string(),
    ))
}

// ---------------------------------------------------------------------------
// Tests — pure helpers against inline S2 shapes, plus the network-free
// resolve_doi guards (empty / strips-to-empty DOI).
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    // ---- strip_doi_url ----
    #[test]
    fn strips_doi_url_prefixes_and_whitespace() {
        assert_eq!(strip_doi_url("https://doi.org/10.1000/xyz"), "10.1000/xyz");
        assert_eq!(strip_doi_url("http://doi.org/10.1000/xyz"), "10.1000/xyz");
        assert_eq!(
            strip_doi_url("https://dx.doi.org/10.1000/xyz"),
            "10.1000/xyz"
        );
        assert_eq!(strip_doi_url("10.1000/xyz"), "10.1000/xyz");
        assert_eq!(strip_doi_url("  10.1000/xyz  "), "10.1000/xyz");
    }

    // ---- is_ratelimited ----
    #[test]
    fn ratelimit_detects_429_only() {
        assert!(is_ratelimited(&CoreError::Upstream(
            "arXiv returned 429 — rate limited; retry in 60s".into()
        )));
        assert!(!is_ratelimited(&CoreError::Upstream(
            "HTTP 404 Not Found".into()
        )));
        assert!(!is_ratelimited(&CoreError::Upstream(
            "Connection reset".into()
        )));
    }

    // ---- arxiv_doi_id ----
    #[test]
    fn arxiv_doi_id_matches_extracts_and_rejects() {
        assert_eq!(
            arxiv_doi_id("10.48550/arXiv.2204.12985").as_deref(),
            Some("2204.12985")
        );
        // case-insensitive prefix.
        assert_eq!(
            arxiv_doi_id("10.48550/ARXIV.2204.12985").as_deref(),
            Some("2204.12985")
        );
        // old-style category/number id.
        assert_eq!(
            arxiv_doi_id("10.48550/arXiv.hep-th/9901001").as_deref(),
            Some("hep-th/9901001")
        );
        // a 5th minor digit is greedily taken (\d{4,5}).
        assert_eq!(
            arxiv_doi_id("10.48550/arXiv.2101.00001").as_deref(),
            Some("2101.00001")
        );
        // non-arXiv DOI and blank -> no match.
        assert!(arxiv_doi_id("10.1000/xyz123").is_none());
        assert!(arxiv_doi_id("").is_none());
    }

    // ---- parse_s2 ----
    #[test]
    fn s2_builds_metadata_without_arxiv_id() {
        let data: Value = serde_json::json!({
            "title": "A Test Paper",
            "authors": [{"name": "Jane Doe"}, {"name": ""}],
            "publicationDate": "2023-05-01",
            "abstract": "An abstract.",
            "externalIds": {},
            "url": "https://www.semanticscholar.org/paper/abc",
        });
        let m = parse_s2(&data, "10.1000/xyz").expect("title present");
        assert_eq!(m.title, "A Test Paper");
        assert_eq!(m.authors, vec!["Jane Doe".to_string()]); // empty name dropped
        assert_eq!(m.source, Some("semanticscholar".into()));
        assert_eq!(m.summary, "An abstract.");
        assert_eq!(m.source_id, "doi:10.1000/xyz");
        assert_eq!(m.doi, Some("10.1000/xyz".into()));
        assert_eq!(
            m.url,
            Some("https://www.semanticscholar.org/paper/abc".into())
        );
        assert_eq!(m.published, NaiveDate::from_ymd_opt(2023, 5, 1).unwrap());
    }

    #[test]
    fn s2_year_fallback_when_no_publication_date() {
        let data: Value = serde_json::json!({
            "title": "Old Paper", "authors": [], "publicationDate": null,
            "year": 2019, "abstract": null, "externalIds": {},
        });
        let m = parse_s2(&data, "10.1000/xyz").unwrap();
        assert_eq!(m.published, NaiveDate::from_ymd_opt(2019, 1, 1).unwrap());
        assert_eq!(m.summary, ""); // null abstract -> ""
    }

    #[test]
    fn s2_invalid_publication_date_falls_back_to_year() {
        let data: Value = serde_json::json!({
            "title": "t", "year": 2018, "publicationDate": "not-a-date",
        });
        let m = parse_s2(&data, "d").unwrap();
        assert_eq!(m.published, NaiveDate::from_ymd_opt(2018, 1, 1).unwrap());
    }

    #[test]
    fn s2_today_when_no_date_or_year() {
        let data: Value = serde_json::json!({
            "title": "Undated Paper", "authors": [], "publicationDate": null,
            "year": null, "abstract": null, "externalIds": {},
        });
        let m = parse_s2(&data, "10.1000/xyz").unwrap();
        assert_eq!(m.published, Utc::now().date_naive());
        // No url/paperId -> the semanticscholar.org/paper/ fallback (empty id).
        assert_eq!(m.url, Some("https://www.semanticscholar.org/paper/".into()));
    }

    #[test]
    fn s2_no_title_returns_none() {
        assert!(parse_s2(&serde_json::json!({"authors": []}), "10.1000/xyz").is_none());
        assert!(parse_s2(&serde_json::json!({}), "d").is_none());
    }

    // ---- resolve_doi network-free guards ----
    #[tokio::test]
    async fn resolve_rejects_empty_doi() {
        let dir = tempfile::tempdir().unwrap();
        let err = resolve_doi("", dir.path(), "")
            .await
            .expect_err("empty DOI must error");
        assert_eq!(err.http_status(), 400);
        assert!(err.to_string().contains("Please enter a DOI"), "got {err}");
    }

    #[tokio::test]
    async fn resolve_rejects_doi_url_that_strips_to_empty() {
        let dir = tempfile::tempdir().unwrap();
        let err = resolve_doi("https://doi.org/", dir.path(), "")
            .await
            .expect_err("a bare doi.org URL strips to empty");
        assert!(err.to_string().contains("Please enter a DOI"), "got {err}");
    }
}
